internal/app/singularity/registry.go

```diff
@@ -0,0 +1,135 @@
+package singularity
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sylabs/scs-library-client/client"
+	scs "github.com/sylabs/scs-library-client/client"
+	"github.com/sylabs/singularity/internal/pkg/client/cache"
+	"github.com/sylabs/singularity/internal/pkg/library"
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+	"github.com/sylabs/singularity/internal/pkg/util/fs"
+	"github.com/sylabs/singularity/internal/pkg/util/uri"
+	"github.com/sylabs/singularity/pkg/signing"
+)
+
+var errNotInCache = fmt.Errorf("image was not found in cache")
+
+// Library is a Registry implementation for Sylabs Cloud Library.
+type Library struct {
+	keystoreURI string
+	arch        string
+
+	client *scs.Client
+	cache  *cache.Handle
+}
+
+// NewLibrary initializes and returns new Library ready to  be used.
+func NewLibrary(scsConfig *scs.Config, cache *cache.Handle, keystoreURI, arch string) (*Library, error) {
```

Comment:
I'm not in favor of `arch` being associated with the `Library` struct I'm afraid. When we use `arch` currently we are asking for a specific architecture on a *single interaction* with the Library. The Library itself supports all architectures - so conceptually we could interact with it multiple times, with each request asking for a different architecture.

I would suggest that the architecture should be specified on the `Pull` request. In this way any currently unimplemented functionality that might want to handle multiple architectures in one invocation is simpler.

Reply:
See @dctrud I wasn't the only one thinking this way :)

After your explanation, I do agree that library and arch are two disjoint parameters to a pull request.

Reply:
Yes, that was my concern as well. My points are:
1. I am not sure other registries will need that, and `Pull` is likely to become and interface method. Thus I didn't want to force other implementations to have/respect arch parameter.
2. A `Library` object is used to perform a single pull. If it was ever reused - no doubt that should have been a `Pull` parameter.

Reply:
If you suspect it will be useful for other registries, then yes, it is reasonable to make it a `Pull` parameter.

Reply:
If the pull request API is going to be an interface, would it make sense to create an opaque "filter spec" (map) argument where the caller can specify arbitrary key-value pairs for filtering requests?

Reply:
Not a fan of that approach. 
Furthermore, it is really straightforward here, we don't have complex filters (only arch now), so I suggest not to complicate things before it is necessary.

Reply:
And it is going to be an interface to enable unit testing of internal CLI package, not a generic purpose interface we are talking about.

Reply:
> If you suspect it will be useful for other registries, then yes, it is reasonable to make it a Pull parameter.

Yes - for docker hub etc. where you can have defined archs. a straight pull into a SIF for another architecture may be useful... especially thinking about deploying to air-gapped / bandwidth constrained systems.

Reply:
Well, currently Docker is smart enough to pull correct arch, unless we want to support corner case where we pull incompatible arch.

Reply:
I am fine with moving arch into `Pull` parameters. Just want to figure out whether that is indeed better :)

Reply:
> Well, currently Docker is smart enough to pull correct arch, unless we want to support corner case where we pull incompatible arch.

Exactly this is what we are doing in the library. Behind the scenes the OCI code we use is handling passing the arch up on the pull request for docker. You can request a different arch. For the library case here we are trying to cater to people who want to be able to pull an `arm` image on their `amd64` system, and then copy it across to their `arm` system. If you consider remote build possibly supporting multiple architectures too then we clearly have to be able to handle interactions which are 'incompatible' with the host arch.

Reply:
okay, will move arch to pull function parameters.
Either way we can change it later as we proceed with this approach with other registries.